### PR TITLE
update venv scripts

### PR DESCRIPTION
**Pull request by thehesiod**
ensures compiled requirements file always match sources used to generate it

### CODE REVIEW

**Comment by thehesiod** [comment]
@jakob-keller thoughts?  this uses another aio-libs module which seems to do what we want

File: build_and_sync_requirements.sh
Comment (thehesiod) [inline build_and_sync_requirements.sh:1]: should make local debugging less painful

**Comment by thehesiod** [comment]
for some reason I can't add you as a reviewer

**Comment by jakob-keller** [comment]
> for some reason I can't add you as a reviewer

I guess that's because I'm not a collaborator.

**Comment by jakob-keller** [comment]
> @jakob-keller thoughts? this uses another aio-libs module which seems to do what we want

Good idea. I did not know about that library.

Do we understand the performance implications? AFAIK, S3 Express is all about lower latencies. It would be a shame, if that cache added more overhead than necessary. A benchmark would be great.

As a user, I would be slightly concerned about adding another dependency: Fewer is better, IMO. In particular, if there is another fairly straightforward solution available as #1073.

**Comment by jakob-keller** [comment]
> Do we understand the performance implications? AFAIK, S3 Express is all about lower latencies. It would be a shame, if that cache added more overhead than necessary. A benchmark would be great.

Here's a simple benchmark.py:

```Python
import asyncio
import functools

import async_lru


async def main(acache, n: int = 1000) -> None:
    """Async main coroutine."""
    await asyncio.gather(*(acache("test") for _ in range(n)))


async def noop(_arg: str) -> None:
    """Trivial noop cooroutine."""


@functools.lru_cache(maxsize=100)
def _cache_1073(_arg: str) -> asyncio.Task[None]:
    return asyncio.create_task(noop(_arg))


async def cache_1073(_arg: str) -> None:
    """functools.lru_cache wraps asyncio.Tasks - #1073."""
    return await _cache_1073(_arg)


@async_lru.alru_cache(maxsize=100)
async def cache_1074(_arg: str) -> None:
    """async-lru based - #1074."""
    return await noop(_arg)
```

Run with

`python -m timeit -s "import asyncio; from benchmark import main, cache_1073" "asyncio.run(main(cache_1073))"`

vs.

`python -m timeit -s "import asyncio; from benchmark import main, cache_1074" "asyncio.run(main(cache_1074))"`

Under CPython 3.11.7 on macOS, #1073 turns out to be slightly faster than #1074: ~10%. That is not significant IMO and should not determine the decision.

**Comment by thehesiod** [comment]
@jakob-keller I don't think #1073 is correct, it's avoiding a needed lock to ensure each lru entry is only awaited once

**Comment by thehesiod** [comment]
IMO any extra effort should be put into speeding up/correcting https://github.com/aio-libs/async-lru. Looking at it I think there are many potential options. external deps aren't great indeed, but at least it's from the same ecosystem (aiolibs)

**Comment by jakob-keller** [comment]
> @jakob-keller I don't think #1073 is correct, it's avoiding a needed lock to ensure each lru entry is only awaited once

I'm still convinced that both approaches are valid fixes for #1072: see https://github.com/aio-libs/aiobotocore/pull/1073#discussion_r1456336296

File: aiobotocore/utils.py
Comment (jakob-keller) [inline aiobotocore/utils.py:1]: I like the fact that this code looks almost the same as upstream.

File: install-requires.txt
Comment (jakob-keller) [inline install-requires.txt:1]: I'm in favor of aligning our "piping" to that of upstream as much as possible: less divergence should result in less trouble down the road.

In this case, botocore inlines its requirements: https://github.com/boto/botocore/blob/1.31.68/setup.py

What is the benefit of spinning this out?

File: setup.py [hunk 1]
```diff
@@ -1,38 +1,32 @@
...
-install_requires = [
-    # pegged to also match items in `extras_require`
-    'botocore>=1.33.2,<1.33.14',
-    'aiohttp>=3.7.4.post0,<4.0.0',
-    'wrapt>=1.10.10, <2.0.0',
-    'aioitertools>=0.5.1,<1.0.0',
-]
+_this_dir = Path(__file__).parent
 
+# NOTE: When updating botocore make sure to update install-requires.txt
```
Comment (jakob-keller) [inline setup.py:7]: We would need to also update documentation elsewhere.

Relying on simultaneous updates to multiple files is also more error prone IMO.

File: setup.py [hunk 2]
```diff
@@ -1,38 +1,32 @@
...
 extras_require = {
     'awscli': ['awscli>=1.31.2,<1.31.14'],
     'boto3': ['boto3>=1.33.2,<1.33.14'],
 }
 
 
-def read(f):
-    return open(os.path.join(os.path.dirname(__file__), f)).read().strip()
+def read(file_name: str):
+    return (_this_dir / file_name).read_text().strip()
```
Comment (jakob-keller) [inline setup.py:15]: `pathlib` sure is more idiomatic. And I love type hints :-) Why not annotate the return type (`-> str`), while we're at it?

**Review by jakob-keller — CHANGES_REQUESTED** [review]
Thank you for the PR and all its contents. I have added a few comments.

IMO, the "piping" related changes should be in another PR. Would you mind splitting the PR up?

**Comment by jakob-keller** [comment]
> > for some reason I can't add you as a reviewer
> 
> I guess that's because I'm not a collaborator.

Now I am. Thank you! 🎉

File: install-requires.txt
Comment (thehesiod) [reply install-requires.txt:1]: the benefit is that we can create a lock file based on the hash of the sources, so we know that the lockfile matches the source.  I suppose we can just hash the whole setup.py.  I can swap to that

File: aiobotocore/utils.py
Comment (thehesiod) [reply aiobotocore/utils.py:1]: going to revert in favor of your impl based on my updated understanding of how it works

**Comment by thehesiod** [comment]
thanks for all the investigative work and instructiveness!

**Comment by thehesiod** [comment]
@jakob-keller ok re-engineered based on discussions

File: build_and_sync_requirements.sh [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+trap cleanup EXIT SIGINT SIGTERM
+
+scratch_file=$(mktemp)
+
+envsubst < requirements-dev.in > ${scratch_file}
+
+# also hash setup.py as it has the install_requires and extras
+SHA_SUM=$(sha1sum ${scratch_file} setup.py | head -c 40)
```
Comment (jakob-keller) [inline build_and_sync_requirements.sh:23]: sha1sum produces two lines with two items each, which are piped to head which takes the first 40 bytes only. Looking at CI runs, this appears to be fragile and it completely ignores the hash of setup.py.

May I propose this alternative?

`cat ${scratch_file} setup.py | sha1sum | awk '{print $1}'`

1. Concatenate both inputs
2. Pipe into sha1sum
3. Extract first word (the 40 character hash)

Comment (thehesiod) [reply build_and_sync_requirements.sh:23]: 🤦🏼‍♂️ thanks, that is a very odd default behavior of sha1sum.  I'll re-work

File: setup.py (hunk 2 above)
Comment (thehesiod) [reply setup.py]: reverted file

File: setup.py (hunk 1 above)
Comment (thehesiod) [reply setup.py]: reverted file

File: install-requires.txt
Comment (thehesiod) [reply install-requires.txt:1]: reverted file

File: aiobotocore/utils.py
Comment (thehesiod) [reply aiobotocore/utils.py:1]: reverted file

**Pull request closed without merging**